src/community/wps-download/src/main/java/org/geoserver/wps/gs/download/RasterEstimator.java

```diff
@@ -126,25 +128,25 @@ public boolean execute(final ProgressListener progressListener, CoverageInfo cov
                 }
                 return true; // EMPTY Intersection
             }
-            // world to grid transform
-            final AffineTransform2D w2G = (AffineTransform2D) reader.getOriginalGridToWorld(
-                    PixelInCell.CELL_CORNER).inverse();
-            final Geometry rasterGeometry = JTS.transform(roiInNativeCRS_, w2G);
 
             // try to make an estimate of the area we need to read
             // NOTE I use the envelope since in the end I can only pass down
             // a rectangular source region to the ImageIO-Ext reader, but in the end I am only going
             // to read the tile I will need during processing as in this case I am going to perform
             // deferred reads
-            areaRead = rasterGeometry.getEnvelope().getArea();
-            // TODO investigate on improved precision taking into account tiling on raster geometry
+            ReferencedEnvelope refEnvelope = JTS.toEnvelope(roiInNativeCRS_.getEnvelope());
+            scaling = new ScaleToTarget(reader, refEnvelope);
 
+            // TODO investigate on improved precision taking into account tiling on raster geometry
         } else {
             // No ROI, we are trying to read the entire coverage
-            final Rectangle2D originalGridRange = (GridEnvelope2D) reader.getOriginalGridRange();
-            areaRead = originalGridRange.getWidth() * originalGridRange.getHeight();
-
+            scaling = new ScaleToTarget(reader);
```

Comment:
You are doing an actual read and scale to estimate the size of the final coverage. If by any chance we are not doing a deferred load this might take forever.